src/write/unit.rs

```diff
@@ -341,26 +371,40 @@ impl Unit {
             strings,
             &range_lists,
             &loc_lists,
-            &mut unit_refs,
-            debug_info_refs,
         )?;
 
         let length = (w.len() - length_base) as u64;
         w.write_initial_length_at(length_offset, length, self.format())?;
 
         for (offset, entry) in unit_refs {
-            let entry_offset = offsets.entry(entry).0;
-            debug_assert_ne!(entry_offset, 0);
             // This does not need relocation.
             w.write_udata_at(
                 offset.0,
-                (entry_offset - offsets.unit.0) as u64,
+                offsets.unit_offset(entry),
                 self.format().word_size(),
             )?;
         }
 
         Ok(offsets)
     }
+
+    /// Reorder base types to come first so that typed stack operations
+    /// can get their offset.
+    fn reorder_base_types(&mut self) {
+        let root = &self.entries[self.root.index];
+        let mut root_children = Vec::with_capacity(root.children.len());
+        for entry in &root.children {
+            if self.entries[entry.index].tag == constants::DW_TAG_base_type {
+                root_children.push(*entry);
+            }
+        }
+        for entry in &root.children {
+            if self.entries[entry.index].tag != constants::DW_TAG_base_type {
+                root_children.push(*entry);
+            }
+        }
```

Comment:
You can use [`Iterator::partition`](https://doc.rust-lang.org/stable/std/iter/trait.Iterator.html#method.partition).

Reply:
That's not quite the same because I want to reorder them, not create two collections. I could then combine the collections, but I prefer to leave it how it is.